src/Database/QueryCompiler.php

```diff
@@ -250,6 +255,71 @@ protected function _buildJoinPart($parts, $query, $generator)
         return $joins;
     }
 
+    /**
+     *
+     * @param array $parts list of joins to be transformed to string
+     * @param \Cake\Database\Query $query The query that is being compiled
+     * @param \Cake\Database\ValueBinder $generator the placeholder generator to be used in expressions
+     * @return string
+     */
+    protected function _buildJoinPartSmartly($parts, $query, $generator)
+    {
+        $alias = [];
+        $joins = [];
+        $order = [];
+        foreach (array_reverse($parts) as $join) {
+            $subquery = $join['table'] instanceof Query || $join['table'] instanceof QueryExpression;
+            if ($join['table'] instanceof ExpressionInterface) {
+                $join['table'] = $join['table']->sql($generator);
+            }
+
+            if ($subquery) {
+                $join['table'] = '(' . $join['table'] . ')';
+            }
+
+            $part = sprintf(' %s JOIN %s %s', $join['type'], $join['table'], $join['alias']);
+
+            $condition = '';
+            if (isset($join['conditions']) && $join['conditions'] instanceof ExpressionInterface) {
+                $condition = $join['conditions']->sql($generator);
+            }
+
+            $ordered = false;
+            if (preg_match_all('/(^|\s)([\w\d]+)\s*\.|`([^`]+)`\s*\./', $condition, $matches, PREG_SET_ORDER)) {
+                foreach ($matches as $match) {
+                    $identified = end($match);
+                    if (isset($alias[$identified]) && $p = array_search($alias[$identified], $order) !== false) {
+                        $order = array_merge(array_slice($order, 0, $p + 1), [$join['alias']], array_slice($order, $p + 1));
+                        $ordered = true;
+                        break;
+                    }
+                }
+            }
+
+            if (!$ordered) {
+                array_unshift($order, $join['alias']);
+            }
+
+            if (strlen($condition)) {
+                $part .= " ON {$condition}";
+            } else {
+                $part .= ' ON 1 = 1';
+            }
+
+            $joins[$join['alias']] = $part;
+
+            $alias[trim($join['alias'], '`')] = $join['alias'];
+            $alias[trim($join['table'], '`')] = $join['alias'];
+        }
+
+        $part = '';
+        foreach ($order as $alias) {
+            $part .= $joins[$alias];
+        }
+
+        return $part;
```

Comment:
This is a significant amount of code without any tests. Adding new modes increases complexity quite a bit as this mode can be combined with every other possible state. Is there a way to accomplish what you are trying to do without additional modes?

Reply:
I can not imagine any solution that will not break applications already deployed in the real world.

My point is:

$query->innerJoin(A...)->innerJoin(B...)
does not generate the same SQL as:
$query->innerJoin(B...)->innerJoin(A...)

But ...

$query->contain(A...)->innerJoin(B...)
$query->innerJoin(B...)->contain(A...)
Generates the same SQL, always leaving the JOIN part of the 'contain' at the end.

Reply:
Has anyone ever had this difficulty?

The only way I can see not to break the implementations already made, is to create an intelligent ordering system for the JOIN part.

Reply:
Or maybe create some parameter that holds the order of the containsJOINs

Reply:
Have you tried replacing `innerJoin` with `innerJoinWith` as @lorenzo suggested earlier? `innerJoinWith` uses the eagerloader so it should preserve join ordering.

> Has anyone ever had this difficulty?

Yes a very similar issue was reported in #10746 but there are a few solutions to that problem.